bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -349,13 +350,29 @@ public void dispose() {
         unregisterDeviceStatusListener(bridgeId);
         cancelJobs();
         stopWebSocket();
+        OAuthClientService oAuthService = this.oAuthService;
+        if (oAuthService != null) {
+            oAuthService.removeAccessTokenRefreshListener(this);
+            oAuthFactory.ungetOAuthService(thing.getUID().getAsString());
+            this.oAuthService = null;
+        }
         client = null;
         deviceStructMan = null;
 
         super.dispose();
         logger.debug("LIVISI SmartHome bridge handler shut down.");
     }
 
+    @Override
+    public void handleRemoval() {
+        OAuthClientService oAuthService = this.oAuthService;
+        if (oAuthService != null) {
+            oAuthService.removeAccessTokenRefreshListener(this);
```

Comment:
This made me wonder: What is the sequence when deleting the thing: I suppose `dispose()` is also called, but which method is called first?

Reply:
I assume that `handleRemoval` is called `dispose` but @J-N-K or @kaikreuzer could certainly confirm that.

Reply:
I can test it later. The problem could be that either `oAuthService` is already disposed here, or best case we don't need to remove the listener here, because it will be done by dispose. I think we just need to verify/debug/test it.

Reply:
The case of thing removal is documented but we do not have a lifecycle like this one: https://www.openhab.org/docs/developer/bindings/#lifecycle

Reply:
Yes, I also checked that page first. 🙂

Reply:
I found the call to `oAuthService.removeAccessTokenRefreshListener(this)` in `handleRemoval` in the ecobee binding if I correctly remember.

An important word was missing in my message. I guess that handleRemoval is called **beofre** dispose.

Reply:
OK, so in that case you don't need to call `removeAccessTokenRefreshListener` here, since `dispose` will do that.

Reply:
In case dispose is called (just after handleRemoval. I am not sure about that.
In fact, you're right, we need to know the lice cycle of the thing removal to be sure what to do.

Reply:
But as I am trying to clean all bindings using the OH OAuth2 client, it is better to be sure to do it it properly in the 8 bindings using `addAccessTokenRefreshListener`.

Reply:
Here is core code:
```
    @Override
    public void thingRemoving(Thing thing, ThingTrackerEvent thingTrackerEvent) {
        setThingStatus(thing, buildStatusInfo(ThingStatus.REMOVING, ThingStatusDetail.NONE));
        notifyThingHandlerAboutRemoval(thing);
    }

    @Override
    public void thingRemoved(final Thing thing, ThingTrackerEvent thingTrackerEvent) {
        logger.debug("Thing '{}' is no longer tracked by ThingManager.", thing.getUID());

        ThingHandler thingHandler = thingHandlers.get(thing.getUID());
        if (thingHandler != null) {
            final ThingHandlerFactory thingHandlerFactory = findThingHandlerFactory(thing.getThingTypeUID());
            if (thingHandlerFactory != null) {
                unregisterAndDisposeHandler(thingHandlerFactory, thing, thingHandler);
                if (thingTrackerEvent == ThingTrackerEvent.THING_REMOVED) {
                    safeCaller.create(thingHandlerFactory, ThingHandlerFactory.class).build()
                            .removeThing(thing.getUID());
                }
            } else {
                logger.warn("Cannot unregister handler. No handler factory for thing '{}' found.", thing.getUID());
            }
        }

        if (!things.containsKey(thing.getUID())) {
            logger.error("Trying to remove thing '{}', but is not tracked by ThingManager. This is a bug.",
                    thing.getUID());
        } else {
            if (!thing.equals(things.remove(thing.getUID()))) {
                logger.error(
                        "Trying to remove thing '{}', but it is different from the thing with the same UID tracked by ThingManager. This is a bug.",
                        thing.getUID());
            }
        }

        missingPrerequisites.remove(thing.getUID());
    }
```

Looks like `thingRemoved` is the one calling `dispose` while `thingRemoving` calls `handleReoval`.
In normal conditions, I guess that `thingRemoving` is called and then `thingRemoved` just after.

Reply:
So yes, it is probably not necessary to call removeAccessTokenRefreshListener in handleRemoval.
I am removing the call.